drivers/i2c/i2c_esp32.c

```diff
@@ -0,0 +1,781 @@
+#define I2C_ESP32_TIMEOUT \
+	((I2C_ESP32_YIELD_THRESHOLD) + (I2C_ESP32_SPIN_THRESHOLD))
+
+enum i2c_esp32_opcodes {
+	I2C_ESP32_OP_RSTART,
+	I2C_ESP32_OP_WRITE,
+	I2C_ESP32_OP_READ,
+	I2C_ESP32_OP_STOP,
+	I2C_ESP32_OP_END
+};
+
+struct i2c_esp32_cmd {
+	u32_t num_bytes : 8;
+	u32_t ack_en : 1;
+	u32_t ack_exp : 1;
+	u32_t ack_val : 1;
+	u32_t opcode : 3;
+	u32_t : 17;
+	u32_t done : 1;
+};
+
+struct i2c_esp32_data {
+	union dev_config config;
+	u16_t address;
+
+	struct k_sem fifo_sem;
+	struct k_sem transfer_sem;
+};
+
+struct i2c_esp32_config {
+	const struct {
+		volatile u32_t *ctr;
+		volatile u32_t *fifo_cfg;
+		volatile u32_t *fifo;
+		volatile u32_t *slave_addr;
+		volatile u32_t *intr;
+		volatile u32_t *intr_clr;
+		volatile u32_t *intr_status;
+		volatile u32_t *intr_raw;
+		volatile u32_t *status;
+		volatile u32_t *cmd_base;
+		volatile u32_t *timeout;
+		volatile u32_t *scl_low_period;
+		volatile u32_t *scl_high_period;
+		volatile u32_t *scl_start_h_time;
+		volatile u32_t *scl_rstart_s_time;
+		volatile u32_t *scl_stop_h_time;
+		volatile u32_t *scl_stop_s_time;
+		volatile u32_t *sda_hold_time;
+		volatile u32_t *sda_sample_time;
+	} regs;
+
+	const struct {
+		int sda_out;
+		int sda_in;
+		int scl_out;
+		int scl_in;
+	} sig;
+
+	const struct {
+		int scl;
+		int sda;
+		bool scl_pull_up;
+		bool sda_pull_up;
+		bool default_level;
+	} pins;
+
+	const struct {
+		int clk;
+		int rst;
+	} enable_bits;
+
+	const struct {
+		bool tx_lsb_first;
+		bool rx_lsb_first;
+	} mode;
+
+	const struct {
+		int source;
+		int line;
+	} irq;
+
+	const union dev_config default_config;
+};
+
+static void i2c_esp32_set_device_enabled(const struct i2c_esp32_config *config,
+					 bool setting)
+{
+	volatile u32_t *periph_clk_reg = (u32_t *)DPORT_PERIP_CLK_EN_REG;
+	volatile u32_t *periph_rst_reg = (u32_t *)DPORT_PERIP_RST_EN_REG;
+
+	if (setting) {
+		*periph_clk_reg |= config->enable_bits.clk;
+		*periph_rst_reg &= ~config->enable_bits.rst;
+	} else {
+		*periph_clk_reg &= ~config->enable_bits.clk;
+		*periph_rst_reg |= config->enable_bits.rst;
+	}
+}
+
+static const char *i2c_esp32_get_gpio_for_pin(int pin)
+{
+	if (pin < 32) {
+#if defined(CONFIG_GPIO_ESP32_0)
+		return CONFIG_GPIO_ESP32_0_NAME;
+#else
+		return NULL;
+#endif /* CONFIG_GPIO_ESP32_0 */
+	}
+
+#if defined(CONFIG_GPIO_ESP32_1)
+	return CONFIG_GPIO_ESP32_1_NAME;
+#else
+	return NULL;
+#endif /* CONFIG_GPIO_ESP32_1 */
+}
+
+static int i2c_esp32_configure_pins(int pin, bool pull_up, bool default_level,
+				    int matrix_out, int matrix_in)
+{
+	const char *device_name = i2c_esp32_get_gpio_for_pin(pin);
+	struct device *gpio;
+
+	if (!device_name) {
+		return -EINVAL;
+	}
+	gpio = device_get_binding(device_name);
+	if (!gpio) {
+		return -EINVAL;
+	}
+
+	gpio_pin_configure(gpio, pin,
+			   GPIO_DIR_OUT | GPIO_DS_DISCONNECT_LOW |
+			   pull_up ? GPIO_PUD_PULL_UP : GPIO_PUD_NORMAL);
+	gpio_pin_write(gpio, pin, default_level);
+
+	gpio_matrix_out(pin, matrix_out, 0, 0);
```

Comment:
Given the naming is ambiguous and generic: probably worth a comment explaining that these calls are to the ESP-32 ROM-resident library layer and not part of this driver or Zephyr's gpio stack.

Or maybe we want to do some abstracting of our own and rename these symbols for our own use with esp32_rom_* prefixes or something.  Not really relevant to this patch.